components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,89 +22,250 @@ import android.os.ParcelFileDescriptor
 import android.provider.MediaStore
 import androidx.annotation.VisibleForTesting
 import androidx.core.app.NotificationManagerCompat
+import androidx.core.content.FileProvider
 import androidx.core.net.toUri
 import androidx.localbroadcastmanager.content.LocalBroadcastManager
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers.IO
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
 import mozilla.components.browser.state.state.content.DownloadState
 import mozilla.components.concept.fetch.Client
 import mozilla.components.concept.fetch.Header
 import mozilla.components.concept.fetch.Headers.Names.CONTENT_LENGTH
 import mozilla.components.concept.fetch.Headers.Names.CONTENT_TYPE
 import mozilla.components.concept.fetch.Headers.Names.REFERRER
+import mozilla.components.concept.fetch.MutableHeaders
 import mozilla.components.concept.fetch.Request
 import mozilla.components.concept.fetch.toMutableHeaders
 import mozilla.components.feature.downloads.ext.addCompletedDownload
 import mozilla.components.feature.downloads.ext.getDownloadExtra
 import mozilla.components.feature.downloads.ext.withResponse
-import mozilla.components.support.base.ids.NotificationIds
+import mozilla.components.support.base.ids.cancel
 import mozilla.components.support.base.ids.notify
 import java.io.File
 import java.io.FileOutputStream
 import java.io.IOException
+import java.io.InputStream
 import java.io.OutputStream
+import kotlin.random.Random
 
 /**
  * Service that performs downloads through a fetch [Client] rather than through the native
  * Android download manager.
  *
  * To use this service, you must create a subclass in your application and it to the manifest.
  */
-abstract class AbstractFetchDownloadService : CoroutineService() {
+@Suppress("TooManyFunctions", "LargeClass")
+abstract class AbstractFetchDownloadService : Service() {
 
     protected abstract val httpClient: Client
     @VisibleForTesting
     internal val broadcastManager by lazy { LocalBroadcastManager.getInstance(this) }
     @VisibleForTesting
     internal val context: Context get() = this
 
-    override fun onCreate() {
```

Comment:
`onCreate` and `startForeground` are needed for Android to count this as a foreground service.

Reply:
I'm confused what you mean by this. It seems to work properly without these functions being overriden. Can you point me to where you found that? :)

Reply:
Without `startForeground`, the service is counted as a background service as is more likely to get killed by the system if it needs memory. https://developer.android.com/reference/android/app/Service

Using `Context.startForegroundService` enforces that `startForeground` is called, but I opted not to use it so that the fallback is being a background service rather than crashing.
https://medium.com/pongploydev/why-start-the-service-on-android-o-and-encounter-a-problem-record-context-startforegroundservice-3c8ddd8969ef

Reply:
I am using `startForeground`, though. Just not immediately. Is this really a problem?

Reply:
Oops I haven't found it, can you link to where it's called? 😅